### ISSUE TITLE
CPU speed comparison among KittenTTS, Piper, MatchaTTS, and Kokoro TTS

### ISSUE THREAD

**Issue by csukuangfj** [KittenML/KittenTTS#40]
I have added a C++ runtime for KittenTTS and compared its RTF with Piper, Matcha, and Kokoro.

The result is in https://github.com/k2-fsa/sherpa-onnx/pull/2460

<img width="939" height="831" alt="Image" src="https://github.com/user-attachments/assets/1247c28f-83e3-4da4-96dc-f8d1f7b4b67a" />

Compared to Kokoro, KittenTTS is faster and smaller in file size.

But when it is compared to Piper, KittenTTS is **SLOW**. 

Also, if the piper tts model is quantized into int8, KittenTTS's model file size is slightly larger  (23MB vs 22 MB)

Has anyone also benchmarked the speed of KittenTTS and compared it with other TTS models on CPU?

**Comment by UsernamesLame** [comment]
Did you run all the tests with a single thread like you ran KittenTTS?

**Comment by csukuangfj** [comment]
> Is the KittenTTS support streamable inference?

@JohnHerry 
 
No.

**Comment by ktomanek** [comment]
I am finding KittenTTS also to be much slower than Piper. Tested on Mac and Raspberry Pi 5:

https://github.com/ktomanek/edge_tts_comparison?tab=readme-ov-file#performance

Was hoping Kitten could be a replacement for Piper on Raspberry Pi, but really can't beat it in terms of speed right now (audio quality much better though).

**Comment by LucaBenini** [comment]
This benchmark is fundamentally flawed.

**1. Wrong model**
The benchmark doesn’t even use the released model. It’s been downgraded from:

```onnx
opset_import {
  version: 20
}
opset_import {
  domain: "ai.onnx.ml"
  version: 5
}
```

to:

```onnx
opset_import {
  version: 19
}
opset_import {
  domain: "ai.onnx.ml"
  version: 4
}
```

This isn’t a trivial change: it can alter runtime behavior and performance. Was this done to run on an outdated ONNX Runtime? If so, you’re not measuring what you think you are.

---

**2. Misuse of RTF**
RTF (Real-Time Factor) is **time\_elapsed / time\_generated**. For a tts model, that means strictly **String → audio** inference time.

The code here doesn’t follow that. It includes *one-time initialization overhead* like:

* ONNX Runtime setup
* Model load from disk
* Session creation

This is not inference time. By that metric, a fast model even a top-tier GPU model would be “slow” if it’s large enough to take time to load from disk. That’s not how you benchmark inference performance.

---

**3. Thread settings cripple parallel models**
Setting both inter- and intra-threads to `1` artificially cripples models that can scale with parallelism.

If you insist on `num_threads == 1`, you must also set **`ORT_SEQUENTIAL`**  otherwise you’re just adding `ORT_PARALLEL` overhead for no reason.

Also, Sherpa’s `num_threads` handling is misleading:

```cpp
sess_opts.SetIntraOpNumThreads(num_threads);
sess_opts.SetInterOpNumThreads(num_threads);
```

This means `num_threads × num_threads` total threads, if the model can use them. These two values exist separately because different models need different tuning forcing them equal hides real performance potential.

---

**Conclusion:**
You’re not benchmarking inference speed. You’re measuring a Frankenstein combination of model loading, environment setup, and crippled runtime settings.

**Comment by csukuangfj** [comment]
@LucaBenini

By the way, I hope you can benchmark the kitten tts model on your own and see how fast it is on CPU, on any CPU you have, not limited to CPU on embedded systems, e.g., cortex A72 or cortex A55, etc.

I am not the only person who has benchmarked kitten tts on CPU.

---

## Benchmark by others

@ktomanek 
https://github.com/ktomanek/edge_tts_comparison?tab=readme-ov-file#performance

<img width="945" height="531" alt="Image" src="https://github.com/user-attachments/assets/5adde292-e7a2-4df0-a46c-9d4b657352fd" />


## From discord

<img width="981" height="661" alt="Image" src="https://github.com/user-attachments/assets/661c80cf-bdcc-4b8a-ae02-baff26eae05e" />

**Comment by ktomanek** [comment]
Interesting, @LucaBenini . I'll take a closer look at your version later and will profile which part in my code slows it down. I am using espeak-ng and the py phonemizer lib. Would you mind resharing your measurements, especially  in comparison to Kokoro (or piper).

**Comment by csukuangfj** [comment]
> You litterally posted the discord screenshot where I told you that for is running above realtime an a raspberry pi and surely faster than Kokoro.

I’ve already shared benchmark results comparing KittenTTS and Kokoro in this issue, which also show that KittenTTS is faster than Kokoro.

> In C on Raspberry PI (5) after the initialization it takes 3 seconds to generate 5 seconds of audio 

So the RTF for your result is 3/5 = 0.6, which agrees with my result.

By the way, from your benchmark code,
https://github.com/LucaBenini/CKittenTTS
you never set explicitly how many threads to use, so it uses all of your CPUs during running. You can use `top` or `htop` to verify that.

To summarize your benchmark result on Raspberry Pi 5:

 - It uses all 4 Cortex A76 CPUs to get an RTF at 0.6

It is faster than Kokoro, but **significantly.slower than Matcha TTS and Piper TTS**. I have published [benchmark results](https://k2-fsa.github.io/sherpa/onnx/tts/pretrained_models/rtf.html) for Piper TTS and Matcha TTS models on the Raspberry Pi 4B, which is less powerful than the Raspberry Pi 5 (Cortex-A72 vs. Cortex-A76).

---

> No it is not. A model can have a set of ops that can be parallized and another not. And for each of these ops one op can be parallelized or not depending of the op choosen.

I agree with your point. However, my goal is to benchmark all models using only a single CPU thread. Since you have a different perspective, I think it would be very helpful if you could share your RTF results under various thread counts, so people can better understand how KittenTTS performs.

The main idea is to give users a clear picture of KittenTTS’s speed across different scenarios.

Not everyone wants to dedicate all of their device’s CPU cores to running KittenTTS — many users need to leave some cores available for other tasks.


> Again the same problem: phonemize_espeak inside is going to call (via the piper dll) espeak_ng_SetVoiceByName that is loading the "en" voice from the disk each time.

No, that’s not correct. The benchmark only measures neural network inference time. eSpeak (or any text frontend), audio I/O, and other overheads are not included. I’ve re-posted the figure below to make this explicit.

<img width="1724" height="1018" alt="Image" src="https://github.com/user-attachments/assets/c185ea8a-3d96-456a-ab5a-699dbf674199" />


> The colab is clearly showing to you that something is wrong in the benchmark: the output ranges from 0.9 to 0.6 a 30% variation in the measures of the time, could be hardware due to the disk access in cloud, could be something else, but obviusly 30% variation means the measures is not good.

It clearly shows how KittenTTS performs across multiple runs. If you feel it is not good, please share your benchmark results across multiple runs. You have already write the C code, so it should not be difficult to get such numbers.

**Comment by csukuangfj** [comment]
Please reopen it.

**Comment by csukuangfj** [comment]
> The correct structure of a benchmark is what @ktomanek has done, with a warmup phase a multiple repetition.

In my Colab, I have a for loop that runs 8 times, and the RTF is printed for each iteration. How many warmup steps do you think are sufficient?

Also, neither of your implementations explicitly sets the number of threads, so they end up using all the device’s CPUs to run a single model.

**Comment by LucaBenini** [comment]
Updating the benchmark with the new 0.2 model: on the raspberry I see a performance decrease of 8% (it is 8% slower than 0.1) on PC the performance decrease is approx 4%. The time went from 3.28 to 3.55.

**Comment by utkarshshukla2912** [comment]
First of all loved the thread thanks for doing such a detailed comparison. I wanted to know the distilled (pun intended) outcome, i understand matcha TTS is the fastest. I also wanted to know how you folks would compare MatchTTS vs Piper in terms of quality.
